ambari-server/src/main/java/org/apache/ambari/server/topology/BlueprintBasedClusterProvisionRequest.java

```diff
@@ -51,13 +51,15 @@
   public BlueprintBasedClusterProvisionRequest(AmbariContext ambariContext, SecurityConfigurationFactory securityConfigurationFactory, Blueprint blueprint, ProvisionRequest request) {
     this.blueprint = blueprint;
     this.request = request;
-
     stackIds = ImmutableSet.copyOf(Sets.union(blueprint.getStackIds(), request.getStackIds()));
-    stack = ambariContext.composeStacks(stackIds);
     mpacks = ImmutableSet.<MpackInstance>builder().
       addAll(blueprint.getMpacks()).
       addAll(request.getMpacks()).build();
 
+    ambariContext.downloadMissingMpacks(mpacks);
+
```

Comment:
I am a bit reluctant to put such side effects into constructors. I think TopologyManager.provisionCluster() was a better place for this.

Reply:
I agree.  But the download needs to be performed during replay, too, in case mpacks were deleted in the meantime.  I'm not sure that's a scenario Ambari needs to support, we can defer that for now.